mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java

```diff
@@ -259,17 +259,47 @@ public boolean isModSeqPermanent(MailboxSession session) {
         if (!isWriteable(mailboxSession)) {
             throw new ReadOnlyException(getMailboxPath(), mailboxSession.getPathDelimiter());
         }
-        Map<MessageUid, MessageMetaData> uids = deleteMarkedInMailbox(set, mailboxSession);
+
+        List<MessageUid> uids = retrieveMessagesMarkedForDeletion(set, mailboxSession);
+        Map<MessageUid, MessageMetaData> deletedMessages = deleteMessages(uids, mailboxSession);
 
         eventBus.dispatch(EventFactory.expunged()
             .randomEventId()
             .mailboxSession(mailboxSession)
             .mailbox(getMailboxEntity())
-            .metaData(ImmutableSortedMap.copyOf(uids))
+            .metaData(ImmutableSortedMap.copyOf(deletedMessages))
             .build(),
             new MailboxIdRegistrationKey(mailbox.getMailboxId()))
             .block();
-        return uids.keySet().iterator();
+        return deletedMessages.keySet().iterator();
+    }
+
+    private List<MessageUid> retrieveMessagesMarkedForDeletion(final MessageRange messageRange, MailboxSession session) throws MailboxException {
+        final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+
+        return messageMapper.execute(
+            () -> messageMapper.retrieveMessagesMarkedForDeletion(getMailboxEntity(), messageRange));
+    }
+
+    @Override
+    public void delete(List<MessageUid> messageUids, MailboxSession mailboxSession) throws MailboxException {
+        Map<MessageUid, MessageMetaData> deletedMessages = deleteMessages(messageUids, mailboxSession);
+
+        eventBus.dispatch(EventFactory.expunged()
```

Comment:
Isn't that `eventBus dispatch` the exact same compared to the one performed by the expunge?

Can't we extract some common chunks of code?

Reply:
I know, I was just trying to make it work first (thus the draft status for now). That is part of my refactoring list already :)